chall04/osalmine.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/python3
+
+# Python 3.8.2
+
+import sys
+import fileinput
+
+def append_space(list, size):
+    for i, s in enumerate(list):
+        if len(s) == 0:
+            list[i] = "     "
+        elif len(s) < size:
+            amount = size - len(s)
+            for _ in range(amount):
+                s += ' '
+            list[i] = s
+        elif len(s) > size:
+            raise Exception("line '{0}' shouldn't have more then {1} characters".format(s, size))
+    return list
+
+def replace_sand(list, i, j, size):
+    ind = len(list) - i - 1
+    for k, search in enumerate(list[ind:]):
+        if ind + k is size or list[ind + k + 1][j] is not ' ':
+            tmp = list[ind][:j] + ' ' + list[ind][j + 1:]
+            list[ind] = tmp
+            tmp = list[ind + k][:j] + '.' + list[ind + k][j + 1:]
+            list[ind + k] = tmp
+            return list
+
+def main():
+    inp = []
+    for line in fileinput.input():
+        inp.append(line)
+    list = [x[:-1] for x in inp]
+    for str in list[1:]:
+        for char in str:
+            if char != ' ' and char != '#' and char != '.':
+                raise Exception("input can only contain ' ', '#' and '.'")
+    if not list[0].isdigit():
+        raise Exception("size can't be other than digits")
+    size = int(list[0])
```

Comment:
If you'd do that before the previous check, you could also add a size check when you iterate on actual data, rather than doing it in different function (which is not an explicit way to do that)